lib/z80rules.1

```diff
@@ -6889,11 +6889,212 @@ $notcpu 8085
 	ld	de,%1	;const
 	ld	hl,%2	;const
 
+%title Reverse primary/secondary load
 %check 4 <= %2 <= 8
-	ld	hl,%1X
+	ld	hl,%1	;const
 	push	hl
 	call	l_gint%2sp	;
 	pop	de
 =
 	call	l_gint%eval(%2 2-)sp	;
 	ld	de,%1
+
+%title Optimise secondary load (perf)
+	ld	hl,(_%1)	X
+	ld	h,0
+	ex	de,hl
+	ld	hl,%2	;const
+=
+	ld	a,(_%1)
+	ld	e,a
+	ld	d,0
+	ld	hl,%2	;const
+
+%title Use 8 bit addition
+%check 0 <= %2 <= 256
+	ld	hl,(_%1)
+	ld	h,0
+	ld	bc,%2
+	add	hl,bc
+	ld	h,0
+=
+	ld	a,(_%1)
+	add	%eval(%2 256 + 256 %%)
+	ld	l,a
+	ld	h,0
+
+%title Load directly into a rather than via hl
+	ld	hl,_%1
+	ld	a,(hl)
+	pop	de
+	ld	(de),a
+	ld	hl,%2
+=
+	ld	a,(_%1)
+	pop	de
+	ld	(de),a
+	ld	hl,%2
+
+	ld	hl,_%1
+	ld	a,(hl)
+	pop	de
+	ld	(de),a
+	ld	l,a
+=
+	ld	a,(_%1)
+	pop	de
+	ld	(de),a
+	ld	l,a
+
+%notcpu gbz80
+	push	hl
+	ld	a,%1
+	pop	de
```

Comment:
(curious without knowing anything about rules in z88dk):
this ends with both HL and DE equal to some value, while `ex de,hl` will not keep HL... does the compiler never emit code which would exploit copy of HL in DE, so the optimization is safe?

Reply:
A push hl/pop de cycle occurs when storing a value to the stack.

An example of where this comes in is, something simple like this:

```
unsigned char b;

int func()
{
        unsigned char c;

        c = b;
}
```

Which is obviously a noop if you've got a context sensitive optimiser...

At -O0 it's ridiculously verbose:

```
// Make space on stack
        dec     sp
// Calculate &a
        ld      hl,0    ;const
        add     hl,sp
// Save it the we store it
        push    hl
// Load b 
        ld      hl,_b
        ld      l,(hl)
        ld      h,0
        ld      h,0
// Get address of a back
        pop     de
// And store it
        ld      a,l
        ld      (de),a
        inc     sp
        ret
```

Which will now end up as:

```
        dec     sp
        ld      hl,0    ;const
        add     hl,sp
        ex      de,hl
        ld      a,(_b)
        ld      (de),a
// Optimiser is context less, so assumes that the value is needed, so loads into primary reg
        ld      l,a
        ld      h,0
        inc     sp
        ret
```

Which you could take a little further:

```
         ld      a,(_b)
        push    af
        inc     sp
        ld      l,a
        ld      h,0
        inc     sp
        ret
```